fix: support seasons swaps in PBR and TerrainHelper

hakasapl opened the pull request:
This allows the vanilla function to execute and create the vanilla materials and shaders. We technically do not need the materials and shaders to be created for PBR but it allows hooks from other plugins to run.

This enables PBR to see the updated swaps from seasons of skyrim for example. There is a 2nd part of fixing seasons with PBR which is in this PR for seasons of skyrim: https://github.com/powerof3/SeasonsOfSkyrim/pull/7

Let me know if there is a better way to avoid the entire vanilla function running in addition to PBR, but I couldn't think of one where PBR would work with seasons.

Tested with a PBR landscape retexture, seasonal swaps work fine now including special features like glint on the swapped parts.

### Other Attempted Fixes

* Reading the vanilla material directly after it is created to get the swap info. This is problematic because we only have texture names and not actual TextureSet forms so it would be impossible to match with required PBR jsons.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **New Features**
  - Introduced seasonal texture swapping for landscape and terrain materials, enhancing visual variety based on seasons.
- **Refactor**
  - Optimized material setup order to ensure consistent texture application without altering user experience.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

hakasapl committed:
refactor: moved vanilla terrain function call above PBR

alandtse reviewed (changes requested):
Please confirm with test builds.

hakasapl commented:
> Please confirm with test builds.

Definitely, once the seasons PR gets merged it's probably best to test then - what do you think? I can mark this as draft for now

alandtse commented:
There's also the test that nothing breaks with the changes. You don't need seasons for that.

hakasapl committed:
switches to alternate method for swap detection from seasons

hakasapl committed:
Merge branch 'doodlum:dev' into pbr-seasons-fix

alandtse commented:
@hakasapl the bot had some interesting thoughts on code outside your changes that seem related to some of the swapping issues you mentioned. You should take a look and see if it makes sense.

hakasapl committed:
fixed potential nullptr in GetSeasonalSwap
added nodiscard to GetSeasonalSwap
fixed PBR meta not being used in SetupLandscapeTexture

hakasapl committed:
Fixed PBR meta not being used for seasonal swaps

hakasapl committed:
Merge branch 'doodlum:dev' into pbr-seasons-fix

hakasapl committed:
Merge branch 'doodlum:dev' into pbr-seasons-fix

hakasapl committed:
fix(PBR): duplicate textureSet object for seasonal swap

hakasapl commented:
@alandtse this should be ready now, users report test build working

alandtse commented on src/Utils/Game.cpp:
```diff
@@ -182,4 +182,19 @@ namespace Util
+			return nullptr;
+		}
+
+		if (textureSet->pad12C > 0) {
```
Do you know what pad12C is? It may be worth upstreaming.

hakasapl replied:
It is an unused uint32. Recent seasons version puts the formid swap there so other plugins can read it. It's not used by the game currently

alandtse commented on src/Utils/Game.h:
```diff
@@ -77,4 +77,9 @@ namespace Util
 	};
+
+	/**
+	 * Helper functions for seasons
```
Write a real description please. @coderabbitai please suggest an appropriate Docstring.

hakasapl replied:
updated

alandtse reviewed (changes requested):
Very close.

hakasapl committed:
chore(hooks): update docstring for GetSeasonalSwap()

pre-commit-ci[bot] committed:
style: 🎨 apply pre-commit.ci formatting

Automated formatting by clang-format, prettier, and other hooks.
See https://pre-commit.ci for details.

hakasapl commented:
Closes #891
